### PR TITLE
Draft: Building GitHub Pull Requests using Jenkins Pipelines

### PR DESCRIPTION
**Pull request by pascalw**

### CODE REVIEW

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+# Building Github Pull Requests with Jenkins Pipelines
+
+We've written about building Github Pull Requests using Jenkins [before](/building-github-pull-requests-with-jenkins/), back in 2013. Now in 2016 we're still strong Jenkins users here at Kabisa and Jenkins has changed quite a bit in recent years. Most notably Jenkins  introduced the notion of [Pipelines](https://jenkins.io/solutions/pipeline/).
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: ... We've written about building Github Pull Requests [back in 2013](...). Three years later we're still strong Jenkins users here at Kabisa, even though Jenkins changed quite a bit. Most notably it introduced...

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+Jenkins Pipelines allow you to describe your entire CI/CD pipeline in code and version it together with your production code. Additionally the whole Jenkins Pipelines ecosystem allows for many powerful setups, including [building jobs in Docker containers](https://wiki.jenkins-ci.org/display/JENKINS/CloudBees+Docker+Pipeline+Plugin) and inbuilt support for building Github Pull Requests.
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: I think `built-in` is a more commonly used synonym of `inbuilt`

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  |- ...
+```
+
+All GitHub projects will be nested under a GitHub namespace. Jenkins will build all master branches and any Pull Requests.
+
+To get started navigate to the `New Item` page in Jenkins and configure as follows:
+
+![](https://kabisa-website-assets.s3.amazonaws.com/theguild/posts/building-github-pull-requests-with-jenkins-pipeline-style/create-organisation-directory.png)
+
+On the next screen some details about your GitHub origanisation have to be configured:
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: typo: organisation

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+![](https://kabisa-website-assets.s3.amazonaws.com/theguild/posts/building-github-pull-requests-with-jenkins-pipeline-style/configure-github-organisation.png)
+
+* **Owner**: enter the name of your GitHub organisation here
+* **Scan credentials**: select or create credentials that should be used to communicate with the GitHub API. You should use a GitHub OAuth token here that has permissions to access all repositories you'd like to build in Jenkins.
+* **Checkout credentials**: Configures which credentials should be used to clone repositories. By default https will be used using the same OAuth token used for scanning but you can also chose to use an SSH key here.
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: ...By default https will be used with the same OAuth token used for scanning; Alternatively you can use SSH.

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+Adapt the other fields to your own taste. With the configuration shown in the screenshot Jenkins will build only the `master` branch of each project and all Pull Requests.
+
+Hit the save button and Jenkins will start scanning your GitHub account for repositories containing a `Jenkinsfile` in the project root. `Jenkinsfile`s tell Jenkins how your project should be built. If this is your first interaction with Jenkins Pipelines you probably won't have any projects that have a Jenkinsfile so read on to learn how to create a Jenkinsfile.
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: ... so read on to learn how to create one.

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+The most basic `Jenkinsfile` looks like this:
+
+```groovy
+node {
+    echo 'Hello from Pipeline'
+}
+```
+
+Obviously this doesn't do much useful, but bear with me. Put this in a file named `Jenkinsfile` in the root of a Github project, commit and push. With this your project can now be built by a Jenkins pipeline, however Jenkins doesn't know about this project yet. By default Jenkins will only look for new projects once a day. To ensure the project is recognized immediately you can press the `Re-scan Organization > Run Now` button. Jenkins will then scan the GitHub organisation again and this time find the `Jenkinsfile` we just commited and setup the project for you.
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: ... Jenkins will then scan the GitHub organisation again, this time finding the Jenkinsfile we've just committed and setting up the project for you.

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+## Power of pipelines
+
+This only scratches the surface of what Jenkins Pipelines can do. Please refer to the [resources](#resources) listed below to read up on Jenkins Pipelines.
+
+## Troubleshooting
+
+If you have any issues setting this up please feel free to reply in the comments. Some things to keep in mind:
+
+* The GitHub OAuth token you use at minimal needs *write* access to your projects to be able to set the build status on Pull Requests.
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: The GitHub OAuth token you use needs at least write access to your projects...

File: drafts/building-github-pull-requests-with-jenkins-pipelines.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+* If your OAuth token also provides organisation administration permissions Jenkins will automatically set up required webhooks. If you don't want Jenkins to do all this make sure you configure the webhook yourself. [Details here](https://github.com/jenkinsci/github-organization-folder-plugin#hook-configuration).
```
Comment (ruisalgado) [inline drafts/building-github-pull-requests-with-jenkins-pipelines.md]: ...make sure **to**...

**Comment by pascalw** [comment]
Thanks @ruisalgado!

**Pull request merged**